### ISSUE TITLE
[macOS] Support policy changes; Xcode 14 and 16 will be removed from macOS 14 on November 4

### ISSUE THREAD

**Issue by erik-bershel** [actions/runner-images#10703]
### Breaking changes

Current support policy:

```
- all OS compatible versions side-by-side
- for beta, GM versions - latest beta only
- old patch versions are deprecated in 3 months
```

New support policy:

```
- only one major version of Xcode will be supported per macOS version
- all minor versions of the supported major version will be available
- beta and RC versions will be provided "as-is" in the latest available macOS image only no matter of beta/GA status of the image
- when a new patch version is released, the previous patch version will be replaced
```

- What does it means in general terms? Starting from the specified date, all minor versions of `Xcode 14` and `Xcode 16` will be removed from the `macOS-14` image.
- What about other images? These changes will not affect `macOS-12` and `macOS-13` images. The changes will affect `macOS-14` and all subsequent images. As a result, `macOS-14`-based images will contain only `Xcode 15.x` versions, `macOS-15` images will contain only `Xcode 16.x` versions, and so on.
- What happens when a new major `Xcode` `beta` is released and there is no next revision of the `macOS` image? Such a `beta` will be shipped as part of the latest `macOS` image that supports this `Xcode`, until the next base `macOS` image is released, for example: `Xcode 16` was shipped as part of `macOS-14` until we released `macOS-15`.

### Target date

November 4, 2024

### The motivation for the changes

Recently, mainly due to the introduction of the new `visionOS` platform, `macOS`-based images have grown to such sizes that they are becoming more difficult to maintain, and users are experiencing difficulties with available disk space, which is critical for completing builds. Parent issue:
- https://github.com/actions/runner-images/issues/10511

### Possible impact

Workflows based on `macOS-14` will stop running if they depend on `Xcode 14` and/or `Xcode 16`

### Platforms affected

- [X] Azure DevOps
- [X] GitHub Actions

### Runner images affected

- [ ] Ubuntu 20.04
- [ ] Ubuntu 22.04
- [ ] Ubuntu 24.04
- [ ] macOS 12
- [ ] macOS 13
- [ ] macOS 13 Arm64
- [X] macOS 14
- [X] macOS 14 Arm64
- [X] macOS 15
- [X] macOS 15 Arm64
- [ ] Windows Server 2019
- [ ] Windows Server 2022

### Mitigation ways

Users whose workflows depend on `Xcode 14` and/or `Xcode 16` will be forced to use the `macOS-13` image for `Xcode-14` and the `macOS-15` image for `Xcode-16`, respectively. Be aware that `macos-13` YAML label refer to Intel-based image.

**Comment by Cheesebaron** [comment]
This is a bit messed up since `macos-15` is not available in Azure DevOps Pipelines. Will they be available in time for this change?

**Comment by freak4pc** [comment]
Won't this basically break any iOS/Swift pipeline that depends on macOS-14 out there? What's the remediation path here? 
Moving to macOS-15 forces Xcode 16 to my understanding.

**Comment by LotskiAndrej** [comment]
> This is a bit messed up since `macos-15` is not available in Azure DevOps Pipelines. Will they be available in time for this change?

Can we get a response on this?

**Comment by erik-bershel** [comment]
Hey @Cheesebaron and @LotskiAndrej!

We are working on adding macOS-15 to ADO, but no ETA or promises are available now. I'll publish info in https://github.com/actions/runner-images/issues/10686 once available.

**Comment by mjosipovic-tgr** [comment]
Is there any way to use Xcode 16 currently using macOS-14 ? I run task for checking Xcode version on ADO and got this, but it says it cannot locate xcode 16
Xcode.app
Xcode_14.3.1.app
Xcode_14.3.app
Xcode_15.0.1.app
Xcode_15.0.app
Xcode_15.1.0.app
Xcode_15.1.app
Xcode_15.2.0.app
Xcode_15.2.app
Xcode_15.3.0.app
Xcode_15.3.app
Xcode_15.4.0.app
Xcode_15.4.app
Xcode_16.0.0.app
Xcode_16.0.app
Xcode_16.1.0.app
Xcode_16.1.app
Xcode_16.1_beta_2.app
Xcode_16.app

**Comment by erik-bershel** [comment]
> Is there any way to use Xcode 16 currently using macOS-14 ?

Yes, we have not removed `Xcode 16` from `macOS-14`. It will be done on October 28.

> I run task for checking Xcode version on ADO and got this, but it says it cannot locate xcode 16

Recommended task: https://learn.microsoft.com/en-us/azure/devops/pipelines/ecosystems/xcode?view=azure-devops

**Comment by mjosipovic-tgr** [comment]
Thanks for the response, but since I'm using MAUI, I cannot add xcWorkspacePath, and ADO crashes if I don't enter anything

**Comment by mjosipovic-tgr** [comment]
I tried it few times, changed the order of the tasks in yaml, and managed to get it to work, tnx for the responses.

**Comment by bengavin** [comment]
> Hey @Cheesebaron and @LotskiAndrej!
> 
> We are working on adding macOS-15 to ADO, but no ETA or promises are available now. I'll publish info in #10686 once available.

We have numerous customers with production MAUI applications that are targeting the Apple App Store, actively being built with Xcode 16, as is currently supported [and 'required' by the current .NET 8 MAUI workloads].  Many of those same customers pay for ADO hosted pipeline time so they can successfully build, test and deploy their applications.  Removing this support before Azure DevOps has an available macos-15 alternative feels short sighted at best.  Doing so with less than 1 month notice will be hard to explain to our customers that rely on this functionality.

I understand the difficulty of managing all the various Xcode versions, but the answer cannot be that we leave customers without viable options.

**Comment by Cheesebaron** [comment]
> 👋 We are shifting our plans for the week due to a number of circumstances.

Thanks that gives us a bit more time to figure out how to tackle this issue, whether to run on our own hardware or move from Azure DevOps Pipelines.

**Comment by erik-bershel** [comment]
Hey @Cheesebaron, `macos-15` image with `Xcode 16` will be available on `Azure DevOps` before the deprecation anyway.

**Comment by erik-bershel** [comment]
`macOS-15` is available using `Azure DevOps` 🚀 
More info: https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml

**Comment by bengavin** [comment]
> `macOS-15` is available using `Azure DevOps` 🚀 More info: https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml

Thank you to the team!

**Comment by Pastajello** [comment]
Some of the tooling for migrating from macOS-14 to 15 is inaccessible and first should be fixed:
(https://github.com/actions/runner-images/pull/10856)

**Comment by mjosipovic-tgr** [comment]
We changed to macOS-15 vM, but we are having issues with android build

Starting: Build Android MAUI app
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.246.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
Script contents:
.
.
.
========================== Starting Command Output ===========================
.
.
.
.


/Users/runner/hostedtoolcache/dotnet/packs/Microsoft.Android.Sdk.Darwin/34.0.143/targets/Microsoft.Android.Sdk.Tooling.targets(31,5): error XA5205: Cannot find `aapt`. Please install the Android SDK Build-Tools package with the `/Users/runner/Library/Android/sdk/tools/android` program. 

Workload updates are available. Run `dotnet workload list` for more information.

##[error]Bash exited with code '1'.
Finishing: Build Android MAUI app

Is this a known issue, on macOS-14 it worked ok, but iOS stopped working on macOS-14
ILLINK : error MT2362: The linker step 'ClassHandleRewriter' failed during processing: One or more errors occurred. (The type 'UIKit.UITextFormattingViewController' (used as a parameter in UIKit.UITextView/_UITextViewDelegate.DidBeginFormatting) is not available in iOS 17.5 (it was introduced in iOS 18.0). Please build with a newer iOS SDK (usually done by using the most recent version of Xcode). 

a lot of these errors were popping up

**Comment by thisisthekap** [comment]
@mjosipovic-tgr Were you able to solve your issue? I am asking because I am facing the same issue currently.

**Comment by bengavin** [comment]
> @mjosipovic-tgr Were you able to solve your issue? I am asking because I am facing the same issue currently.

We just do our Android builds on macos-14, which continues to work fine :)

**Comment by zjhch123** [comment]
> `macOS-15` is available using `Azure DevOps` 🚀 More info: https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml

Hi @erik-bershel , Is there any plan to support macOS-15-arm in ADO?

**Comment by erik-bershel** [comment]
Hey @zjhch123!

Not at the moment as I know. Updates will be available here: https://github.com/actions/runner-images/issues/8971
List of available images and related updates also are available to track [here](https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml).

**Comment by ADKaster** [comment]
@erik-bershel It would be nice if in the future blog posts like this one included ALL breaking changes upcoming. https://github.blog/changelog/2024-11-05-notice-of-breaking-changes-for-github-actions/

Since this change was not in the blog post, we were blind sided by this change, as we don't check the runner-images repository regularly for CI that has continued to work for a long time.

**Comment by Saadnajmi** [comment]
With this change, I no longer have an image that has both Xcode 16 and visionOS (needs an ARM machine) with Azure Pipelines. That breaks our workflows heavily.

**Comment by KingkorAtMaxint** [comment]
Is this a joke? Why would anyone remove something in the same image version introducing a breaking change in the same version? Doesn't this completely destroys the meaning of having a version system in the first place?

This is breaking all of our iOS build + release pipelines.

**Comment by isnine** [comment]
Looking at the PR's that have been mentioned in this issue and the thumbs down emoticons from people, a very large number of people don't like this change and is blocked by this change.

I understand that lack of disk space is an issue, but at least it can be solved simply by `rm xcode16.0.app`. However, removing the pre-installed Xcode 16.0 would require users to spend a long time manually downloading it themselves.

@erik-bershel  So can we think about revert this change? **Or at least we can consider this change after ADO provides MacOS-15 Arm64?**

**Comment by mjosipovic-tgr** [comment]
> @mjosipovic-tgr Were you able to solve your issue? I am asking because I am facing the same issue currently.

@thisisthekap , Yes, I added new task into the yaml file 

`  - task: Bash@3
    
displayName: "Install Android SDK and Build Tools"
    inputs:   targetType: "inline"
      script: | 
       export ANDROID_HOME=$HOME/Library/Android/sdk
       export PATH=$ANDROID_HOME/cmdline-tools/latest/bin:$ANDROID_HOME/build-tools/34.0.0:$ANDROID_HOME/platform-tools:$PATH
        mkdir -p $ANDROID_HOME/cmdline-tools/latest
        mkdir -p $ANDROID_HOME/build-tools
        sdkmanager --install "cmdline-tools;latest" "platform-tools" "build-tools;34.0.0" "platforms;android-34"
        yes | sdkmanager --licenses`

and for the iOS certificate issue, in task for installing certificates I added this

` opensslPkcsArgs: "-legacy"
            continueOnError: true`
            
 Now both android and iOS works on macOS-15

**Comment by thisisthekap** [comment]
@mjosipovic-tgr Is there any specific reason why you added `continueOnError: true` for iOS?

**Comment by mjosipovic-tgr** [comment]
> @mjosipovic-tgr Is there any specific reason why you added `continueOnError: true` for iOS?

@thisisthekap ah yes, sorry, I have tasks for both SSL compatibility, in case one of them fails, it will continue and won't stop the build. Depending what you are using you can use only one of them, if you using old certificate and new version of OpenSSL you can remove the one without legacy flag, vice versa. Hope this helps

```
 # Two tasks for handling OpenSSL compatibility
          - task: InstallAppleCertificate@2
            inputs:
              certSecureFile: nameOfFile.p.12
              keychain: "temp"
              certPwd: yourPassword
              deleteCert: true
              opensslPkcsArgs: "-legacy"
            continueOnError: true
            displayName: "Install Apple Certificate with Legacy Flag"

          - task: InstallAppleCertificate@2
            inputs:
              certSecureFile: nameOfFile.p.12
              keychain: "temp"
              certPwd: yourPassword
              deleteCert: true
            continueOnError: true
            displayName: "Install Apple Certificate without Legacy Flag"
```

**Comment by thisisthekap** [comment]
@mjosipovic-tgr Thank you for that! Thats a perfect workaround for the mess with different image versions used randomly we are currently facing.

**Comment by tsal** [comment]
> Looking at the PR's that have been mentioned in this issue and the thumbs down emoticons from people, a very large number of people don't like this change and is blocked by this change.
> 

This change has completely disrupted our release process, and it was with zero notice. This PR doesn't satisfy any kind of "notice".

And to rub salt in that, we've been getting billed for the minimum time, every. time. we. restart.

Why? because this has been intermittently failing our builds, but today it's been near-every-time.

Since this wasn't any kind of notification, we assumed it was something wrong with our builds, then later chalked up to "that terrible github experience" - turns out we were correct, just the reasons for the terrible experience were outright choices on GH's end.

Fix this; revert it and give us time.

**Comment by charlesroddie** [comment]
It makes no sense to remove new software from a system (XCode 16) while preserving older software (XCode 15), unless there are known problems with the newer software. If you want to keep just one version, keep XCode 16.

This change forced us to migrate to `macos-15`, but this doesn't work because of https://github.com/microsoft/azure-pipelines-tasks/issues/20628#issuecomment-2472701880 .

**Comment by ddramowicz** [comment]
Can you please clarify what the final decision is on this? It appears that this was implemented in this PR https://github.com/actions/runner-images/pull/10834 and then reverted again yesterday in this PR https://github.com/actions/runner-images/pull/10962 ...

**Comment by erik-bershel** [comment]
Hey there! 👋 

Due to a combination of factors (including some unexpected issues with dependencies on `Xcode 16`), we've decided to mitigate the issue by reverting `Xcode 16` to `macOS-14` images. We'll be working on the notification model and testing, and will be back with more details on upcoming changes and plans.

**Comment by dsmitchell** [comment]
Please keep Xcode 16. Thanks to Xamarin.UITest we need to build IPA's with Xcode 16 and run the tests with Xcode 15.4

**Comment by yasali** [comment]
> Hey there! 👋
> 
> Due to a combination of factors (including some unexpected issues with dependencies on `Xcode 16`), we've decided to mitigate the issue by reverting `Xcode 16` to `macOS-14` images. We'll be working on the notification model and testing, and will be back with more details on upcoming changes and plans.

When we will be able to see the changes, Xcode 16 is still not available with macos-14-xlarge runner?

**Comment by mjosipovic-tgr** [comment]
> Hey there! 👋
> 
> Due to a combination of factors (including some unexpected issues with dependencies on `Xcode 16`), we've decided to mitigate the issue by reverting `Xcode 16` to `macOS-14` images. We'll be working on the notification model and testing, and will be back with more details on upcoming changes and plans.

@erik-bershel  I hope you don't remove it from macOS-15 since I managed to got it to work.

**Comment by yasali** [comment]
> > Hey there! 👋
> > Due to a combination of factors (including some unexpected issues with dependencies on `Xcode 16`), we've decided to mitigate the issue by reverting `Xcode 16` to `macOS-14` images. We'll be working on the notification model and testing, and will be back with more details on upcoming changes and plans.
> 
> @erik-bershel I hope you don't remove it from macOS-15 since I managed to got it to work.

Which Xcode version it works for you in macOS-15?

**Comment by mjosipovic-tgr** [comment]
@yasali xcode 16

**Comment by tkhho** [comment]
> Hey there! 👋
> 
> Due to a combination of factors (including some unexpected issues with dependencies on `Xcode 16`), we've decided to mitigate the issue by reverting `Xcode 16` to `macOS-14` images. We'll be working on the notification model and testing, and will be back with more details on upcoming changes and plans.

@erik-bershel Will this happen on Azure DevOps as well? We have there the same issues like on GitHub Runners, but still don't have the Xcode 16 on mac os 14 again. Our builds are blocked right now.

**Comment by Utsira** [comment]
This change means it is not impossible to build against iOS 17 and iOS 18. Can the iOS 17 Simulator be added to the macOS-15 image please?

**Comment by yeftha** [comment]
> > > Hey there! 👋
> > > Due to a combination of factors (including some unexpected issues with dependencies on `Xcode 16`), we've decided to mitigate the issue by reverting `Xcode 16` to `macOS-14` images. We'll be working on the notification model and testing, and will be back with more details on upcoming changes and plans.
> > 
> > 
> > @erik-bershel I hope you don't remove it from macOS-15 since I managed to got it to work.
> 
> Which Xcode version it works for you in macOS-15?

@yasali , what fix you have given to make xcode16 work on macOS-15. Can you please help me.

**Comment by yasali** [comment]
> > > > Hey there! 👋
> > > > Due to a combination of factors (including some unexpected issues with dependencies on `Xcode 16`), we've decided to mitigate the issue by reverting `Xcode 16` to `macOS-14` images. We'll be working on the notification model and testing, and will be back with more details on upcoming changes and plans.
> > > 
> > > 
> > > @erik-bershel I hope you don't remove it from macOS-15 since I managed to got it to work.
> > 
> > 
> > Which Xcode version it works for you in macOS-15?
> 
> @yasali , what fix you have given to make xcode16 work on macOS-15. Can you please help me.
We are using Xcode 16.1 with macOS-15, it works well now.

**Comment by erik-bershel** [comment]
Hey @yasali and @yeftha!

I will be very grateful if you share the problems you have when migrating from `macos-14` to `macos-15` or simply when using the `macos-15` image. This can be done either as a separate issue or [here](https://github.com/actions/runner-images/issues/10686). Sooner or later the `macos-15` image will be the latest and we would like to resolve all possible problems by then.

**Comment by yeftha** [comment]
> Hey @yasali and @yeftha!
> 
> I will be very grateful if you share the problems you have when migrating from `macos-14` to `macos-15` or simply when using the `macos-15` image. This can be done either as a separate issue or [here](https://github.com/actions/runner-images/issues/10686). Sooner or later the `macos-15` image will be the latest and we would like to resolve all possible problems by then.

@erik-bershel , Thank you for the quick response, since morning I faced issues with the pipeline (macos-15,xcode16.1) for installApplecertificate.

**Error message:**
Error outputting keys and certificates
40AB4351F87F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
Could not find certificate from <stdin>
##[warning]Error parsing certificate. This might be caused by an unsupported algorithm. If you're using old certificate with a new OpenSSL version try to set -legacy flag in opensslPkcsArgs input.
##[error]Error: /usr/local/bin/openssl failed with return code: 1
Finishing: InstallAppleCertificate

I have just tried running the pipeline and it worked fine without any issues. Thank you

**Comment by yeftha** [comment]
> Hey @yasali and @yeftha!
> 
> I will be very grateful if you share the problems you have when migrating from `macos-14` to `macos-15` or simply when using the `macos-15` image. This can be done either as a separate issue or [here](https://github.com/actions/runner-images/issues/10686). Sooner or later the `macos-15` image will be the latest and we would like to resolve all possible problems by then.

Hi @erik-bershel ,

I am running the pipeline in Azure DevOps, using xcode16.1 with macOS-15, getting below error at install apple certificate task. 

**Error message**
Error outputting keys and certificates
40AB5345F87F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:355:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
Could not find certificate from <stdin>
##[warning]Error parsing certificate. This might be caused by an unsupported algorithm. If you're using old certificate with a new OpenSSL version try to set -legacy flag in opensslPkcsArgs input.
##[error]Error: /usr/local/bin/openssl failed with return code: 1
Finishing: InstallAppleCertificate

can you please help me on this

**Comment by MerSpyro** [comment]
@yeftha For this you can check th solution provided by mjosipovic-tgr in 
 https://github.com/actions/runner-images/issues/10703#issuecomment-2472826230

**Comment by erik-bershel** [comment]
Heads up! ⏰ 

To provide immediate mitigation, we have decided to roll back the current `macos-14` images in production to the previous stable version from November 22. This should take about `24 hours`, but may take up to `48 hours` for some specific cases that can be worked around with a restart.

**Comment by yeftha** [comment]
> @yeftha For this you can check th solution provided by mjosipovic-tgr in [#10703 (comment)](https://github.com/actions/runner-images/issues/10703#issuecomment-2472826230)

@MerSpyro This fix is working fine . Thank you

**Comment by woutermeuwis** [comment]
XCode 16 still does not seem available in the macOS-14 image?
When will this be available again?

**Comment by erik-bershel** [comment]
@woutermeuwis please switch to `Xcode 16` before running your workflow. Issue states that the current Xcode version is set to 15.4, but it does not mean that there is no `Xcode 16` at all.
Example action: https://github.com/maxim-lobanov/setup-xcode

**Comment by erik-bershel** [comment]
👋 

For those tracking these changes or encountering issues for the first time, here’s some context and clarification.


As you may know, we recently faced significant challenges where many users began running out of disk space during tests and builds on `macOS`-based runners. 😞  This issue was triggered by the inclusion of `visionOS` simulators in the latest `Xcode` builds. The root cause, however, lies in our policy of providing nearly all supported minor versions of `Xcode` on each `macOS` image, leading to steadily increasing disk usage as `Apple`’s ecosystem grows.

Here’s how we’ve approached resolving the issue so far:
- **Immediate action:** We temporarily stopped pre-installing `visionOS` platform tools in `Xcode` on `macOS-14` images.
- **Strategic adjustment:** We introduced a policy to limit the number of major `Xcode` versions installed per `macOS` image, alongside reinstating `visionOS` tools for supported `Xcode` versions. This aimed to free up significant disk space. https://github.com/actions/runner-images/pull/10876

However, we’ve discovered that some projects rely on multiple major versions of `Xcode` simultaneously, making these strict measures overly restrictive. This prompted us to refine our strategy, prioritising functionality alongside disk space optimisation.

### Current Strategy 📏 

Our updated approach balances reducing installed `Xcode` versions with maintaining runner usability:

1. **`macOS-14` images:**

- Will include all minor releases of `Xcode 15` with the full platform tools suite (as before).
- Will now also include two minor releases of `Xcode 16` (excluding `visionOS` tools). These will follow a “last two” principle, with the oldest replaced by the newest as updates are released. `Beta` versions will not be included. https://github.com/actions/runner-images/pull/10962

2. **`macOS-15` images:**

- Will include all minor releases of `Xcode 16` with the full platform tools suite (as before).
- Will also include the latest available release of `Xcode 15` with the full platform tools suite. https://github.com/actions/runner-images/pull/11023

3. **Disk Space and Future Adjustments:**

- Currently, `macOS-14` images have approximately `60 GB` of free space. However, **this is not a guaranteed figure** and may decrease over time.
- When free space approaches `30 GB`, we will reassess our policies and toolsets to ensure compliance with **our promised minimum** of `14 GB` of free space.
- `macOS-15` images are not currently impacted by this issue. However, with future `Xcode` updates, policies may need to be revisited. 🤷‍♂️ 

#### Recommendations 🦮 

- If your workflows depend on `visionOS`, we recommend switching to `macOS-15` images.
- If your tasks require significant disk space, consider preparing the runner by removing unnecessary large components before starting your workflows.

**Comment by mikeknack86** [comment]
```swift

```

**Comment by amit-thapa** [comment]
@erik-bershel is the macos-15 for the ado now stable? I did an Android build which uses Android API level 34 on the macos-15 and got the following error below. This runs fine on the macos-14

`error XA5207: Could not find android.jar for API level 34. This means the Android SDK platform for API level 34 is not installed; it was expected to be in`

**Comment by erik-bershel** [comment]
Hey @amit-thapa! `macOS-15` is expected to be fine. I think your issue related to the fact that the first available platform on `macOS-15` is 35. You need to install it manually or to switch to the next. See the difference here:
[macOS-14](https://github.com/actions/runner-images/blob/main/images/macos/macos-14-Readme.md#android),
[macOS-15](https://github.com/actions/runner-images/blob/main/images/macos/macos-15-Readme.md#android).

**Comment by sroyaa** [comment]
Hello guys,

Since yesterday we have some issues in our pipelines from this pull request : 
https://github.com/actions/runner-images/pull/11315

The folder Xcode_16 was first deleted than readded in macos 14 and now you guys removed it once again.. 

Can someone provide us advices to deal with change like this ? 
Thanks a lot in advance :)

**Comment by alexlogvin** [comment]
Please, update the docs about macos-14 software https://github.com/actions/runner-images/blob/main/images/macos/macos-14-Readme.md

**Comment by erik-bershel** [comment]
Hey @sroyaa and @alexlogvin!

More info could be find here: https://github.com/actions/runner-images/issues/11203
Updates are in progress here: https://github.com/actions/runner-images/pull/11315

**Comment by mkhtradm01** [comment]
> Hello guys,
> 
> Since yesterday we have some issues in our pipelines from this pull request : #11315
> 
> The folder Xcode_16 was first deleted than readded in macos 14 and now you guys removed it once again..
> 
> Can someone provide us advices to deal with change like this ? Thanks a lot in advance :)

I just got into same issue too 💔

**Comment by erik-bershel** [comment]
We can no longer support all available versions of all `Xcode` within a single `macOS-14` image due to limited disk space. Accordingly, `Xcode 16` will be updated periodically by replacing the oldest with the newly released one. You can track this moment through announcements in the repository or simply switch to `macOS-15` - all supported releases of `Xcode 16` will be available there.

**Comment by sroyaa** [comment]
Hello @erik-bershel,

Thanks a lot for your answer, very clear! :) 

In November we tried to update to MacOS-15  but got issues in Azure Agent allocation as well, hope it is fixed now, we will try this.
Thanks :)

**Comment by Adam-Langley** [comment]
Hi @erik-bershel 
I've just attempted a push to the Apple App Store with our build output, and was rejected because it was built with Xcode 15.4
I've read through the above thread. Thank you for the explanations.

What doesn't make sense (to me), however, is the agent image "macos-latest" points to macos-14, not macos-15.
The MS docs clearly state macos-latest=macos-14, and macos-14 is in preview:
https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops&tabs=yaml

The thing is - I would expect that using the "macos-latest" alias, should be a tool to stay current with requirements to produce production releases - not support historical SDKs, or pre-releases.

As of April 24, 2025, Xcode 16 was required to produce builds compatible with the App Store.
https://developer.apple.com/news/upcoming-requirements/?id=02212025a

Meaning that anyone using "macos-latest" would produce non-compliant output.

I would love my pipelines to stay in compliant with Apple App Store requirements. If "macos-latest" is not the way to achieve this, what is?

Perhaps we need an alias, such as "macos-xcode-latest", which would point to the macos with the latest xcode?

Thank you

**Comment by erik-bershel** [comment]
Hey @Adam-Langley!

Thank you for the input. ❤ 

This is a semantic difference. Terminology varies between ecosystems. In the GitHub/ADO context, the `-latest` in the tag name means not so much access to the very latest features, but the newest of the stable ones. Each ecosystem has its own understanding in this regard. We can, of course, migrate to macOS-15 as quickly as it was added to the ecosystem, but this will only lead to a huge number of problems. 

Regarding the Xcode situation, I should note that macOS images **always contain more than one Xcode instance**, and if you want, you can use Xcode 16.1 or Xcode 16.2 on macOS-14 runners - you just need to add a step to manage the environment configuration, which will not take more than five seconds.

Example script:
```bash
sudo xcode-select --switch /Applications/Xcode_${XCODE_VERSION}.app/Contents/Developer
```

List of Xcode instances on macOS-14 image: [link](https://github.com/actions/runner-images/blob/main/images/macos/macos-14-Readme.md#xcode).

At the moment macOS-14 based images are compliant with Apple App Store requirements as I know. We'll migrate `latest` tag once something changed in that area.

**Comment by Adam-Langley** [comment]
Hi @erik-bershel,

Thank you for your detailed response.
I see the semantic difference - I agree it has it's purposes.

I should probably prefix my comments with: I'm not pushing for "macos-15" to be the "macos-latest", but that there should be an alias for a build agent that will produce app store compliant builds. 
I see that xcode 16 has been reintroduced to macos-14 (based on your previous statement that *macOS-14-based images will contain only Xcode 15.x*). However, the "default" xcode instance on macos-14 is still xcode 15.4 - which is still *noncompliant*.

**As of April 24, 2025, Xcode 16 was required to produce builds compatible with the App Store.
https://developer.apple.com/news/upcoming-requirements/?id=02212025a**

So assuming Microsoft's "app store compliance" intent remains the same - I feel that something should be done.
I guess I'd like to provoke a discussion to answer what that action could be.

I know from our perspective, our goal is to ensure that build output will *always* be compliant with Apple App Store policies. That means, for example, not having to periodically update our build scripts to override defaults on a build agent image.
It's good to read your statement recognising a need from MS to achieve this ("...images are compliant with Apple App Store requirements") - but I haven't seen that written anywhere before. I think this should be promoted to a wiki page, along with exactly what this means. Of course, please direct me to one if it already exists.

I'd like to circle back to my suggestion of this semantic definition of "latest" - I think there is loose-to-no connection between "latest" and "compliant". If I use the build-agent "macos-latest" it may technically contain a compliant xcode image now, but I need to edit my build scripts to forcibly use xcode 16.x vs the default 15.4. Why would 15.x be considered a suitable default after April 24 , 2025?

An easy solution - make sure that the default xcode will produce App Store compliant builds. That means the agent team stays on top of which xcode version is required to publish to the app store, and ensure that is the "default" xcode, on the "default" macos.
I think that's the minimally invasive approach. Can we achieve that?
